src/fps_hooks.rs

```diff
@@ -60,54 +60,76 @@ pub fn get_player_rotate_speed_rate_hook(method_info: OptionalMethod) -> f32 {
+    _padding3: [u8; 2],
+    m_speed: f32,
+    m_blend: f32,
+}
 
+static mut HUB_MOVE_STATE_MOVE_CURRENT_FRAMETIMING: f32 = 2.0;
 
-// NPC Handling
+// floating point pattern helper
+fn fpp_helper(float: f32) -> i32 {
+    return (float * 1000.0) as i32;
+}
 
-// #[skyline::hook(offset = 0x23D3920)]
-// fn hub_move_state_move_spline_loop_ctor(
-//     this: *const u8,
-//     unit: *const u8,
-//     data: &Vector3<f32>,
-//     body_anim: Il2CppString,
-//     face_anim: Il2CppString,
-//     is_turn: bool,
-//     speed: f32,
-//     method_info: OptionalMethod,
-// ) {
-//     let modified_speed = speed_modifier() * speed;
-//     println!("I AM RUNNING!!!!!!!!!!!");
-
-//     call_original!(this, unit, data, body_anim, face_anim, is_turn, modified_speed, method_info);
-// }
-
-
-
-// #[repr(C)]
-// pub struct AppHubMoveStateMoveO {
-//     _padding: [u8; 0x3C], // offset
-//     m_speed: f32,
-// }
-
-// // App.HubMoveStateMove$$IsEnd    71028bfae0    bool App.HubMoveStateMove$$IsEnd(App_HubMoveStateMove_o * __this, MethodInfo * method)    268
-// #[unity::hook("App", "HubMoveStateMove", "IsEnd")]
-// pub fn hub_move_state_move_is_end(this: *mut AppHubMoveStateMoveO, method_info: OptionalMethod) -> bool {
-//     println!("I AM end, i am alive too!!!!!!!!!!!");
-//     // unsafe {
-//     //     // dereference the raw pointer to access the struct
-//     //     let this_ref = &mut *this;
-
-//     //     let result = call_original!(this, method_info);
-
-//     //     this_ref.m_speed *= 0.5;
-//     //     result
-//     // }
-//     return call_original!(this, method_info);
-// }
\ No newline at end of file
+#[unity::hook("App", "HubMoveStateMove", "Start")]
+pub fn hub_move_state_move_start(this: &mut AppHubMoveStateMoveO, resume: bool, method_info: OptionalMethod) {
+    let frametiming = get_frametiming();
+    unsafe {
+        // messy initialization
+        if HUB_MOVE_STATE_MOVE_CURRENT_FRAMETIMING == 2.0 {
+            HUB_MOVE_STATE_MOVE_CURRENT_FRAMETIMING = frametiming;
+        }
+
+        // fix transition from 60 to 30. currently broken
+        if (frametiming - HUB_MOVE_STATE_MOVE_CURRENT_FRAMETIMING) == 0.5 {
+            this.m_speed /= 0.5;
+        }
+        HUB_MOVE_STATE_MOVE_CURRENT_FRAMETIMING = frametiming;
+    }
+    
+    if frametiming == 1.0 {
+        if this.m_speed == 0.0030002 {
+            this.m_speed = 0.006;
+        }
+        return call_original!(this, resume, method_info);
+    } else {
+        match fpp_helper(this.m_speed) {
+            50 | 60 => { // NPC Walk
+                this.m_speed *= frametiming;
+            },
+            30 if this.m_is_turn => {
+                this.m_speed *= frametiming;
+            },
+            4 => { // NPC Pool Stop
+                this.m_speed *= frametiming;
+            },
+            6 => { // NPC Pool Stop
+                this.m_speed *= frametiming;  // this one does NOT like being an odd number
+                this.m_speed += 0.0000002;
```

Comment:
what's this tiny delta for?

Reply:
Yes, when it's odd, the pool npc gets stuck